File: pmt-backend/src/main/java/com/codesolutions/pmt_backend/service/ProjectService.java
package com.codesolutions.pmt_backend.service;

import com.codesolutions.pmt_backend.entity.*;
import com.codesolutions.pmt_backend.repository.ProjectMemberRepository;
import com.codesolutions.pmt_backend.repository.ProjectRepository;
import com.codesolutions.pmt_backend.repository.RoleRepository;
import com.codesolutions.pmt_backend.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ProjectService {

    private final ProjectRepository projectRepository;
    private final ProjectMemberRepository projectMemberRepository;
    private final RoleRepository roleRepository;
    private final UserRepository userRepository;

    public ProjectService(ProjectRepository projectRepository, ProjectMemberRepository projectMemberRepository,
                          RoleRepository roleRepository, UserRepository userRepository) {
        this.projectRepository = projectRepository;
        this.projectMemberRepository = projectMemberRepository;
        this.roleRepository = roleRepository;
        this.userRepository = userRepository;
    }

    /**
     * Crée un nouveau projet et associe l'utilisateur créateur comme ADMIN
     */
    @Transactional
    public Project createProject(Project project, Long creatorId) {
        // Sauvegarde du projet
        Project createdProject = projectRepository.save(project);

        // Récupère l'utilisateur par son ID
        User creator = userRepository.findById(creatorId)
                .orElseThrow(() -> new IllegalArgumentException("Utilisateur non trouvé"));

        // Récupère le rôle ADMIN
        Role adminRole = roleRepository.findByName("ADMIN")
                .orElseThrow(() -> new IllegalStateException("Rôle ADMIN introuvable"));

        // Crée l'association ProjectMember et l'enregistre
        ProjectMember projectMember = new ProjectMember(creator, createdProject, adminRole);
        projectMemberRepository.save(projectMember);

        return createdProject;
    }

    /**
     * Récupère la liste de tous les projets.
     */
    public List<Project> getAllProjects() {
        return projectRepository.findAll();
    }

    public Project getProjectById(Long id) {
        return projectRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Projet avec id " + id + " non trouvé"));
    }


    /**
     * Récupère la liste des projets associés à un utilisateur (créés ou auxquels il est membre).
     */
    @Transactional(readOnly = true)
    public List<Project> getProjectsByUser(Long userId) {
        List<ProjectMember> memberships = projectMemberRepository.findByUserId(userId);
        return memberships.stream()
                .map(ProjectMember::getProject)
                .distinct()
                .collect(Collectors.toList());
    }

}

File: pmt-frontend/src/app/project/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Project {
  id?: number;
  name: string;
  description: string;
  startDate: string;
}

export interface ProjectMember {
  id: {
    userId: number;
    projectId: number;
  };
  user: {
    id: number;
    username: string;
    email: string;
  };
  role: {
    id: number;
    name: string;
  };
}

@Injectable({
  providedIn: 'root'
})
export class ProjectService {
  private apiUrl = environment.apiUrl;

  constructor(private http: HttpClient) { }

  // Récupère tous les projets
  getProjects(): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.apiUrl}/api/projects`);
  }

  // Créer un nouveau projet en associant l'utilisateur créateur en tant qu'ADMIN
  createProject(project: Project, creatorId: number): Observable<Project> {
    return this.http.post<Project>(`${this.apiUrl}/api/projects?creatorId=${creatorId}`, project);
  }

  // Récupère les projets associés à un utilisateur
  getUserProjects(userId: number): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.apiUrl}/api/projects/users/${userId}/projects`);
  }

  // Récupère la liste des membres d'un projet
  getProjectMembers(projectId: number): Observable<ProjectMember[]> {
    return this.http.get<ProjectMember[]>(`${this.apiUrl}/api/projects/${projectId}/members`);
  }

  // Récupère les détails d'un projet par son identifiant
  getProjectById(projectId: number): Observable<Project> {
    return this.http.get<Project>(`${this.apiUrl}/api/projects/${projectId}`);
  }

  updateMemberRole(projectId: number, userId: number, role: string): Observable<ProjectMember> {
    return this.http.put<ProjectMember>(`${this.apiUrl}/api/projects/${projectId}/members/${userId}?role=${role}`, {});
  }
}
